File: 2075. decode-the-slanted-ciphertext.java
class Solution {
    public String decodeCiphertext(String encodedText, int rows) {
        if (rows == 0) return "";

        int n = encodedText.length();
        if (n == 0) return "";

        int cols = n / rows;

        // Step 1: Build matrix
        char[][] mat = new char[rows][cols];
        int idx = 0;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                mat[i][j] = encodedText.charAt(idx++);
            }
        }

        // Step 2: Diagonal traversal
        StringBuilder result = new StringBuilder();

        for (int startCol = 0; startCol < cols; startCol++) {
            int i = 0, j = startCol;

            while (i < rows && j < cols) {
                result.append(mat[i][j]);
                i++;
                j++;
            }
        }

        // Step 3: Remove trailing spaces
        int end = result.length() - 1;
        while (end >= 0 && result.charAt(end) == ' ') {
            end--;
        }

        return result.substring(0, end + 1);
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1
        String encodedText1 = "ch   ie   pr";
        int rows1 = 3;
        String output1 = sol.decodeCiphertext(encodedText1, rows1);
        System.out.println("Example 1: encodedText=\"ch   ie   pr\", rows=3");
        System.out.println("Output: \"" + output1 + "\"");
        System.out.println("Expected: \"cipher\"");
        System.out.println();
        
        // Example 2
        String encodedText2 = "iveo    eed   l te   olc";
        int rows2 = 4;
        String output2 = sol.decodeCiphertext(encodedText2, rows2);
        System.out.println("Example 2: encodedText=\"iveo    eed   l te   olc\", rows=4");
        System.out.println("Output: \"" + output2 + "\"");
        System.out.println("Expected: \"i love leetcode\"");
        System.out.println();
        
        // Example 3
        String encodedText3 = "coding";
        int rows3 = 1;
        String output3 = sol.decodeCiphertext(encodedText3, rows3);
        System.out.println("Example 3: encodedText=\"coding\", rows=1");
        System.out.println("Output: \"" + output3 + "\"");
        System.out.println("Expected: \"coding\"");
    }
}

File: 2075. decode-the-slanted-ciphertext.py
class Solution:
    def decodeCiphertext(self, encodedText: str, rows: int) -> str:
        if rows == 0:
            return ""

        n = len(encodedText)
        if n == 0:
            return ""

        cols = n // rows

        # Step 1: Build matrix
        mat = []
        idx = 0
        for i in range(rows):
            row = []
            for j in range(cols):
                row.append(encodedText[idx])
                idx += 1
            mat.append(row)

        # Step 2: Diagonal traversal
        result = []

        for startCol in range(cols):
            i, j = 0, startCol
            while i < rows and j < cols:
                result.append(mat[i][j])
                i += 1
                j += 1

        # Step 3: Remove trailing spaces
        return "".join(result).rstrip()

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    encodedText1 = "ch   ie   pr"
    rows1 = 3
    output1 = sol.decodeCiphertext(encodedText1, rows1)
    print(f"Example 1: encodedText='{encodedText1}', rows={rows1}")
    print(f"Output: '{output1}'")
    print(f"Expected: 'cipher'")
    print()
    
    # Example 2
    encodedText2 = "iveo    eed   l te   olc"
    rows2 = 4
    output2 = sol.decodeCiphertext(encodedText2, rows2)
    print(f"Example 2: encodedText='{encodedText2}', rows={rows2}")
    print(f"Output: '{output2}'")
    print(f"Expected: 'i love leetcode'")
    print()
    
    # Example 3
    encodedText3 = "coding"
    rows3 = 1
    output3 = sol.decodeCiphertext(encodedText3, rows3)
    print(f"Example 3: encodedText='{encodedText3}', rows={rows3}")
    print(f"Output: '{output3}'")
    print(f"Expected: 'coding'")